File: pog-store/core/src/main/java/br/com/pogstore/core/storage/ConcurrencyException.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package br.com.pogstore.core.storage;

import java.util.UUID;

/**
 *
 * @author Vitor Hugo Salgado <vsalgadopb@gmail.com>
 */
public class ConcurrencyException extends Exception {

	private static final long serialVersionUID = 1L;
	private UUID eventSourceId;
	private long eventSourceVersion;

	public ConcurrencyException() {
		super();
	}

	public ConcurrencyException(Throwable cause) {
		super(cause);
	}

	public ConcurrencyException(UUID eventSourceId, long versionToBeSaved) {
		super(
				String.format(
						"There is a newer than %s version of the event source with id %s you are trying to save stored in the event store.",
						versionToBeSaved, eventSourceId));

		this.eventSourceId = eventSourceId;
		this.eventSourceVersion = versionToBeSaved;
	}

	public UUID getEventSourceId() {
		return eventSourceId;
	}

	public long getEventSourceVersion() {
		return eventSourceVersion;
	}

}

File: pog-store/cms/core/PogStore.Cms.Core/Cqrs/Storage/ConcurrencyException.cs
using System;
using System.Runtime.Serialization;

namespace PogStore.Cms.Core.Cqrs.Storage
{
	[Serializable]
	public class ConcurrencyException : Exception
	{
		public Guid EventSourceId { get; private set; }

		public long EventSourceVersion { get; private set; }

		public ConcurrencyException(Guid eventSourceId, long versionToBeSaved)
			: base(string.Format("There is a newer than {0} version of the event source with id {1} you are trying to save stored in the event store.", versionToBeSaved, eventSourceId))
		{
			EventSourceId = eventSourceId;
			EventSourceVersion = versionToBeSaved;
		}

		protected ConcurrencyException(SerializationInfo info, StreamingContext context)
			: base(info, context)
		{
			EventSourceId = (Guid)info.GetValue("EventSourceId", typeof(Guid));
			EventSourceVersion = info.GetInt64("EventSourceVersion");
		}

		public override void GetObjectData(SerializationInfo info, StreamingContext context)
		{
			base.GetObjectData(info, context);
			info.AddValue("EventSourceId", EventSourceId);
			info.AddValue("EventSourceVersion", EventSourceVersion);
		}
	}
}